### PR TITLE
fix: actually use asyncio

### PR DESCRIPTION
**Pull request by laf070810**
It seems that asyncio is not used correctly. Current implementation runs without IO concurrency, which can be seen by adding some timing outputs. 

For example, the `_finish_jobs` function in `scheduler.py` will wait for the `postprocess` of all jobs sequentially, thus the total execution time of `_finish_jobs` is the sum of all `postprocess` execution time, which can be slow when the computation scales up. 

This PR facilitates the actual use of asyncio. By making the changes in this PR, the total execution time of `_finish_jobs` will be in theory the max of all `postprocess` execution times (slightly longer in reality), which can be much shorter than the sum of times.  I can already see a significant speedup in one of my use cases. 

This PR is just a proposal for now. Asyncio usage in other parts of snakemake may have the same issue. We can proceed with a full fix if this proposal is thought to be correct. 

The file diff on github may not be intuitive. The change is simply surrounding the codes with `async with` and replacing `await` with `asyncio.TaskGroup().create_task()` when possible. 

### QC

* [ ] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [ ] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).

## Summary by CodeRabbit

- **New Features**
    - Enhanced concurrency in job post-processing, improving efficiency and responsiveness.
    - Added detailed timing metrics for job scheduling and processing stages.

- **Bug Fixes**
    - Improved error handling during job processing, ensuring better robustness.

- **Documentation**
    - Added logging for performance metrics related to post-processing and scheduling operations.

### CODE REVIEW

**Comment by laf070810** [comment]
Well, some observations still show that it is sequential. I may need further investigation.

File: snakemake/jobs.py [hunk 1]
```diff
@@ -1104,75 +1105,87 @@
...
+                    tg.create_task(self.dag.handle_log(self))
+                if not error:
+                    tg.create_task(
+                        self.dag.check_and_touch_output(
+                            self,
+                            wait=self.dag.workflow.execution_settings.latency_wait,
+                            ignore_missing_output=ignore_missing_output,
+                            # storage not yet handled, just require the local files
+                            wait_for_local=True,
+                        )
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply snakemake/jobs.py:1148]: Do not suggest to add trailing commas in multi-line function calls.
We use black for autoformatting, which handles this for us.

File: snakemake/scheduler.py [hunk 2]
```diff
@@ -326,63 +351,72 @@
...
+                    if not self.workflow.dryrun:
+                        try:
+                            if self.workflow.exec_mode == ExecMode.DEFAULT:
+                                tg.create_task(
+                                    job.postprocess(
+                                        store_in_storage=not self.touch,
+                                        handle_log=True,
+                                        handle_touch=not self.touch,
+                                        ignore_missing_output=self.touch,
+                                    )
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply snakemake/scheduler.py:365]: Do not suggest to add trailing commas in multi-line function calls.
We use black for autoformatting, which handles this for us.

File: snakemake/jobs.py [hunk 3]
```diff
@@ -1110,75 +1111,87 @@ async def postprocess(
...
+                    )
+                self.dag.unshadow_output(self, only_log=error)
```
Comment (johanneskoester) [inline snakemake/jobs.py:1150]: I think this is problematic. The idea of the await was that the check_and_touch happens before the stuff below like the unshadow. If I don't get this wrong here, now the check and touch happens in the background, in theory in parallel to the unshadow?

**Comment by johanneskoester** [comment]
I think I will close this for now because of my concerns. Nevertheless, feel free to reopen or create a new one as a follow up if I missed something here. Thanks a lot for the suggestion, I still really appreciate the discussion!

**Pull request closed without merging**